redesign, upgrade to Bootstrap 4 and Fork Awesome

thetric opened the pull request:
Upgrades to Bootstrap 4. BS4 [dropped Glyphicons](https://getbootstrap.com/docs/4.1/migration/#components) so I've replace them with [~Font Awesome~](https://fontawesome.com/icons) [Fork Awesome](https://github.com/ForkAwesome/Fork-Awesome) (which also contains more icons).

# TODOs:
- [x] apply styling to inputs
- [x] apply styling for Bootstrap's custom forms (radio and check boxes)
- [x] bug(options/nav): "general" stays active after tab switch
- [x] ~redesign "connected databases" with [cards](https://getbootstrap.com/docs/4.1/components/card/)?~ (later, future PR)
- [x] ~refactor(icons): replace font awesome web fonts + css with svg&js (smaller bundle?)~ (use Fo**rk**Awesome instead)
- [x] refactor(icons): replace Fo**nt**Awesome icons with Fo**rk**Awesome (see https://github.com/keepassxreboot/keepassxc-browser/pull/322#issuecomment-433696065)
- [x] use old green color theme (https://github.com/keepassxreboot/keepassxc-browser/pull/322#issuecomment-433695850)

# files:
- [x] `options/options.html`
- [x] `popups/popup.html`
- [x] `popups/popup_httpauth.html`
- [x] `popups/popup_multiple-fields.html`
- [x] `popups/popup_remember.html`
- [x] `options/shortcuts.html`

# bugs:
- [x] options/site-prefs: no warnings displayed if value is invalid
- [ ] options/site prefs: cannot change "Username-only Detection" on freshly added entry, although the changes are reflected to the model after a reload
- [ ] option/general: "Default group for saving new passwords:" displays "Default: null KeePassXC-Browser Passwords" with `null`?

# Previews:

![image](https://user-images.githubusercontent.com/19861998/65797048-807f7680-e16e-11e9-8cfc-d9fa8c58ca98.png)

![image](https://user-images.githubusercontent.com/19861998/65797067-8d9c6580-e16e-11e9-99d2-dab8e4c0f688.png)

![image](https://user-images.githubusercontent.com/19861998/65797093-a147cc00-e16e-11e9-930b-0187bcc60e9b.png)

![image](https://user-images.githubusercontent.com/19861998/65797105-ac9af780-e16e-11e9-9214-dcd803831edd.png)

![image](https://user-images.githubusercontent.com/19861998/65797166-d05e3d80-e16e-11e9-8dd4-b50e95b9a765.png)

thetric commented:
WIP of the "General" tab

![screenshot_2018-10-27_19-47-01](https://user-images.githubusercontent.com/19861998/47607502-21c2f180-da21-11e8-94ce-bd20d60f406c.png)

phoerious commented:
I like it. I've always wanted to give it a slight makeover myself, but never had the time.

droidmonkey commented:
Oh sexy!

varjolintu commented on keepassxc-browser/manifest.json:
```diff
@@ -100,11 +100,5 @@
         "http://*/*",
         "https://api.github.com/"
     ],
-    "applications": {
```
Why did you remove this?

varjolintu commented:
Looks very nice! Btw, are we going to keep the older green color sheme? The new navbar is bright green.

varjolintu commented:
Some things to keep eye on:
https://github.com/twbs/bootstrap/pull/23586
https://forkawesome.github.io/Fork-Awesome/

thetric commented:
This is mostly done (except the navbar bug).

Due to the change the primary color to the old green the 'success' color was redefined as light purple. Does this look okay? (E.g. in the popup after KeepassXC is connected)

sebast889 commented:
Shouldn't the auto-fill warning come under the "Automatically fill in single-credential entries"? In the above screenshot it looks like it is for the Automatically retrieve credentials" option.

Also, can the main header tabs like "General" be a darker green color when it is the active tab (just like it is for the current layout)?

thetric commented:
> Shouldn't the auto-fill warning come under the "Automatically fill in single-credential entries"? In the above screenshot it looks like it is for the Automatically retrieve credentials" option.

Done:

![screenshot_2018-12-22_18-31-55](https://user-images.githubusercontent.com/19861998/50377126-f66f1400-0617-11e9-80ad-b2960d401d4a.png)

> Also, can the main header tabs like "General" be a darker green color when it is the active tab (just like it is for the current layout)?

Or how do you like this: (The nav bar entries should have the same width)

![image](https://user-images.githubusercontent.com/19861998/50377262-3f27cc80-061a-11e9-96fe-ef90bb147118.png)

sebast889 commented:
First one looks better to me, second has too much white. Is it not possible to match the current navbar layout?

thetric commented:
Fixed the active nav item issue ("general" tab stayed active) and refactored the navbar to look like the current one:

![screenshot_2019-01-26_15-32-18](https://user-images.githubusercontent.com/19861998/51788588-f36bc400-217f-11e9-802c-02165966086c.png)

thetric commented:
Todo: remove script tag for redundant web extensions polyfills in Firefox-only page `shortcuts.html`

droidmonkey commented:
Is this ready for review?

thetric commented:
@droidmonkey yes

varjolintu commented:
@thetric Please check my reviews.

Btw, I was also wondering if the increase of the font sizes is actually wanted here. The popup dialog seems quite huge with much of wasted space.

I also advice you to check the popup content using both Firefox and Chromium and see that it looks identical. Now there are some strange differences in the button sizes etc.

thetric commented:
Reduced font size in popups and fixed inconsistent widths across browsers:

Firefox:

![screenshot_2019-02-16_15-02-27](https://user-images.githubusercontent.com/19861998/52900918-c8afe100-31fc-11e9-9b13-7b5635da92bd.png)

Chrome:

![screenshot_2019-02-16_15-02-52](https://user-images.githubusercontent.com/19861998/52900930-ed0bbd80-31fc-11e9-9fd3-bcf2c0b2d1bf.png)


There is still is little difference in the top button bar but this shouldn't be too bad.

sebast889 commented:
Any reason for the color changes in buttons ("Choose custom login fields for this page" and "Redetect login fields")? I feel the current layouts colors are easier to identify (no 2 buttons have the same color and no transparent button that differs from the look of the other buttons).

droidmonkey commented:
Yah I'm not a fan of the transparent button. Also the text should be shortened, something like "Select Login Fields" with a nice icon (mouse hovering over form?)

Move settings over to the far right, eliminate the text, the icon is appropriate. Add text to the lock icon "Lock Databases". There is a lot to be done to this dialog.

thetric commented:
Firefox:

![screenshot_2019-02-23_14-02-23](https://user-images.githubusercontent.com/19861998/53286765-ab868f80-3773-11e9-81ab-0ca5014b4b7c.png)

Chrome:

![screenshot_2019-02-23_14-03-39](https://user-images.githubusercontent.com/19861998/53286789-fd2f1a00-3773-11e9-9a93-e7287d44b15c.png)

phoerious commented:
The identifier should be on its own line. Breaking it into two lines looks rather ugly.

droidmonkey commented:
I really like the new buttons!

thetric commented:
The navbar in the options page looks weird on <991px:

![screenshot_2019-02-23_14-22-55](https://user-images.githubusercontent.com/19861998/53287019-b0990e00-3776-11e9-930e-d73ff4a92035.png)

thetric commented:
> The identifier should be on its own line. Breaking it into two lines looks rather ugly.

@phoerious Done

![screenshot_2019-02-23_14-29-27](https://user-images.githubusercontent.com/19861998/53287068-74b27880-3777-11e9-88eb-ecffc3a4c3a2.png)

thetric commented:
Fixed the responsive behavior of the navbar:

![screenshot_2019-02-23_15-05-53](https://user-images.githubusercontent.com/19861998/53287455-8fd3b700-377c-11e9-810d-17c4094c1489.png)


![screenshot_2019-02-23_15-04-38](https://user-images.githubusercontent.com/19861998/53287435-6dda3480-377c-11e9-966c-b820621ddd37.png)

varjolintu commented on keepassxc-browser/popups/popup.css:
```diff
@@ -1,51 +1,37 @@
 body {
-    font-family: sans-serif;
     overflow-x: hidden;
     background-color: #eee;
```
Maybe the popup background color should be white, just like the Settings page.

thetric replied:
Yeah, agreed

varjolintu replied:
Nice. Everything else is already looking much better!

thetric commented on keepassxc-browser/options/shortcuts.js:
```diff
@@ -65,53 +65,60 @@ async function updateKeys() {
-    } catch (e) {
-        console.log('Cannot change shortcut: ' + e);
-        createBanner('danger', shortcut);
+        createAlert('success', shortcut);
```
The `alertArea` will be moved to the end of the page.
I'll investigate your bug.

varjolintu commented on build-bootstrap.js:
```diff
@@ -0,0 +1,29 @@
+const inFile = 'keepassxc-browser/styles/bootstrap.scss';
+const outFile = 'keepassxc-browser/styles/bootstrap.css';
+
+renderScss(
```
If we are going to use the build script, the CSS output file should be minimized. Also, I'm not sure if the extension is going to pass the AMO/CWB audit if there's .scss files inside the package.

thetric replied:
The AMO Wiki [doesn't recommend minifying code](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/Source_Code_Submission):

> Obfuscating or minifying your code isn't recommended. It rarely protects your code, as these techniques can be reversed using tools such as JSNice and JS Beautifier. And, unlike the advantage that minified code offers web pages loaded over the internet, as extension code is loaded from a local source the performance benefits are not significant.

[According to the AMO Wiki](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/Source_Code_Submission#Provide_your_extension_source_code) the `bootstrap.scss` file shouldn't present a problem:

> ## When must source code be uploaded?
> You must upload your extension’s source code when its code was created using:
> [...]
> * any other custom tool that takes files, applies pre-processing, and generates file(s) to include in the extension.

However it would be necessary to provide build instructions and [upload the sources when publishing a new addon version](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/Source_Code_Submission#How_to_upload_source_code):

> ## Provide build instructions
> To reproduce the build, the reviewer runs the instructions you provided and then uses a diff tool to compare the generated sources to those in the extension. There must be no differences. The easiest way to provide the build instructions is to include a README file with the submitted source code. [...]
>
> [...]
> When using npm, yarn, or other package management tools that support it, be sure to include the lockfile, for example, `package-lock.json`. Otherwise, reviewers may use a different version resulting in differences between the generated code and that in the extension.
> 
> [...] However, you don’t need to describe how to install common tools such as npm or node.

varjolintu replied:
Still, the .scss file is not needed inside the extension.

thetric commented:
Restyled the popup when multiple logins are available:

![Unbenannt](https://user-images.githubusercontent.com/19861998/54465428-6b676b00-477b-11e9-87ac-e0691aa1adc0.PNG)

varjolintu commented:
FYI, we are going to merge #416 soon. All changes in that PR should be included in this one too. Most of the relevant changes are in the Settings page. And those should be quite easy to copy.

Keep up the good work!

thetric commented:
What needs to be done before merging this? Except integrating the changes from #416

varjolintu commented:
@thetric https://github.com/keepassxreboot/keepassxc-browser/pull/396 would be needed also.

varjolintu commented:
FYI: https://blog.jquery.com/2019/04/10/jquery-3-4-0-released/

But maybe we shouldn't upgrade because: 

> Specifically, jQuery 3.4.0 is deprecating :first, :last, :eq, :even, :odd, :lt, :gt, and :nth. When we remove Sizzle, we’ll replace it with a small wrapper around querySelectorAll, and it would be almost impossible to reimplement these selectors without a larger selector engine.

thetric commented:
How does this relate to this issue? It does not upgrade jQuery

varjolintu commented:
@thetric Yes, it doesn't. It was just FYI, if you were thinking about it.

thetric commented:
Merged the latest changes from `develop`. Ready for review

varjolintu commented on package.json:
```diff
@@ -20,5 +26,10 @@
   },
-  "homepage": "https://github.com/keepassxreboot/keepassxc-browser#readme"
+  "homepage": "https://github.com/keepassxreboot/keepassxc-browser#readme",
+  "browserslist": [
```
Does this follow the minimum versions of Firefox specified in manifest.json?

thetric replied:
Decreased minimum supported Firefox version to 52

varjolintu commented on keepassxc-browser/options/options.html:
```diff
@@ -2,439 +2,1034 @@
+      <h3><img src="/icons/keepassxc_48x48.png" alt="logo" /> KeePassXC-Browser</h3>
+
+      <nav class="navbar navbar-expand-lg navbar-dark bg-primary text-white mb-4">
+        <button
```
Is there any reason for this file to be in different syntax than the rest of the HTML files? Why are the elements divided in multiple lines?

varjolintu commented on keepassxc-browser/options/options.html:
```diff
@@ -2,439 +2,1034 @@
+    <script type="text/javascript" src="../global.js"></script>
+    <script type="text/javascript" src="../jquery-3.3.1.min.js"></script>
+    <script type="text/javascript" src="bootstrap.min.js"></script>
+    <script type="text/javascript" src="options.js"></script>
```
The `type` is not needed with HTML5. See the original file.

varjolintu commented on keepassxc-browser/options/options.html:
```diff
@@ -2,439 +2,1034 @@
+                  type="button"
+                >
+                  <i class="fa fa-floppy-o" aria-hidden="true"></i>
+                  <span data-i18n="optionsButtonSave"></span>
```
Warnings are not displayed when the value is invalid.

varjolintu replied:
Also, why the buttons are at the bottom, not after the input field? Then for example "Default group for saving new passwords" has the save button next to the input.

thetric replied:
The buttons are now placed directly behind the inputs.
The "Blink Time" and "Redirect Offset" options look a bit weird with the `ms` text addon. Should it be removed?

![Screenshot_2019-06-02_18-53-17](https://user-images.githubusercontent.com/19861998/58764500-0811fb80-8568-11e9-8e79-61bdf1ed4ce5.png)

varjolintu replied:
The "ms" is already explained in the help text.

varjolintu commented on keepassxc-browser/options/options.html:
```diff
@@ -2,439 +2,1034 @@
+                minlength="10"
+              />
+              <div class="input-group-append">
+                <button class="btn btn-primary" id="sitePreferencesManualAdd" type="button">
```
Warnings are not displayed when the value is invalid.

thetric commented:
Remaining bugs:

* [ ] no form validation (blink timeout, site prefs > add url)
* [ ] site prefs/table: cannot change "Username-only Detection" on freshly added entry, although the changes are reflected to the model after a reload

varjolintu commented:
We had a discussion and we would like to have this PR in the same release with https://github.com/keepassxreboot/keepassxc-browser/pull/351. The only changes concerning this PR are in the options page where the timeout settings have been removed totally. You should do the same. I'll modify the content script buttons to match the themes in this PR, because we cannot use Bootstrap there.

thetric commented:
The timeout related settings UI + code is removed. The only remaining bug is the appearance of the checkbox in the "Site preferences" tab

thetric commented:
I'm afraid I have no clue why the checkboxes don't toggle correctly. In [my version](https://github.com/thetric/keepassxc-browser/blob/bootstrap-4-upgrade/keepassxc-browser/options/options.js#L314) in the `change` handler of the checkbox the `url` is always `undefined` because `$(this)` refers to the *original* `tr.clone` but not to the *appended row*.

Might somebody point out the failure?

varjolintu commented:
@thetric I'll look into this tomorrow and see if I can find the reason.

varjolintu commented:
#772 replaces this. Closing the PR.
